src/Orleans.Serialization/Serializers/CodecProvider.cs

```diff
@@ -169,39 +169,50 @@ public IFieldCodec TryGetCodec(Type fieldType)
             // If the field type is unavailable, return the void codec which can at least handle references.
             return fieldType is null ? _voidCodec
                 : _untypedCodecs.TryGetValue(fieldType, out var existing) ? existing
-                : TryCreateCodec(fieldType);
-        }
+                : GetOrCreateCodec(fieldType);
 
-        /// <inheritdoc/>
-        public IFieldCodec<TField> GetCodec<TField>()
-        {
-            var res = TryGetCodec<TField>();
-            if (res is null) ThrowCodecNotFound(typeof(TField));
-            return res;
-        }
+            IFieldCodec GetOrCreateCodec(Type fieldType)
+            {
+                var result = TryCreateCodec(fieldType);
+                if (result is not null)
+                {
+                    _untypedCodecs.TryAdd(fieldType, result);
+                }
 
-        private IFieldCodec TryCreateCodec(Type fieldType)
-        {
-            if (!_initialized) Initialize();
+                return result;
+            }
 
-            ThrowIfUnsupportedType(fieldType);
+            IFieldCodec TryCreateCodec(Type fieldType)
```

Comment:
It's better to keep it as a regular method so stack traces look more clear for example.